### PR TITLE
[FREELDR] Re-integrate the ASM files (and corresponding C code) in MSVC builds, that were previously compiled just for GCC builds.

### PR DESCRIPTION
**Pull request by HBelusca**
## Purpose

Do what is exposed in the topic of the PR ;)

## Proposed changes

- Multiboot support added back in FreeLdr MSVC builds so that it can be
  booted using e.g. GRUB. See [CORE-15563](https://jira.reactos.org/browse/CORE-15563).

- Re-introduce the disk drive mapper code.

- Remove the outdated .text16 section merging.

[SDK/INCLUDE] Modify the .code16 macro so that the 16-bit code segment
can get a chance to be correctly merged with other (possibly 32-bit)
code segments, as done e.g. during FreeLdr linking, without keeping
generating multiple .text segments with different attributes and
generating the corresponding LNK4078 warning.

Note that this problem already existed when compiling NTOSKRNL on MSVC:
"v86.S.obj : warning LNK4078: multiple '.text' sections found with different attributes (C0520040)".

## TODO

- [x] Check that I don't break existing multiboot support in the GCC build.
- [x] Perform actual test using GRUB.

### CODE REVIEW

File: boot/freeldr/freeldr/CMakeLists.txt [hunk 1]
```diff
@@ -90,12 +90,14 @@ list(APPEND FREELDR_COMMON_SOURCE
 
 if(ARCH STREQUAL "i386")
     list(APPEND FREELDR_COMMON_ASM_SOURCE
+        arch/i386/drvmap.S
+        # arch/i386/mb.S
```
Comment (HBelusca) [inline boot/freeldr/freeldr/CMakeLists.txt]: It appears that this file is a left-over from good-ol' times and can actually be removed completely from compilation.
Please let me know if you think it should be used somewhere...

File: boot/freeldr/freeldr/CMakeLists.txt [hunk 2]
```diff
@@ -178,12 +177,11 @@ endif()
...
 
 add_library(freeldr_common
+    ${freeldr_common_asm}
     ${FREELDR_BOOTLIB_COMMON_SOURCE}
     ${FREELDR_ARC_COMMON_SOURCE}
     ${FREELDR_NTLDR_COMMON_SOURCE}
-    ${FREELDR_COMMON_SOURCE}
-    ${freeldr_common_asm}
-    linuxboot.c)
+    ${FREELDR_COMMON_SOURCE})
```
Comment (HBelusca) [inline boot/freeldr/freeldr/CMakeLists.txt]: linuxboot.c is actually redundant since it's already in the `FREELDR_COMMON_SOURCE` list.

File: boot/freeldr/freeldr/CMakeLists.txt [hunk 3]
```diff
@@ -199,8 +197,8 @@ set(PCH_SOURCE
 add_pch(freeldr_common include/freeldr.h PCH_SOURCE)
 add_dependencies(freeldr_common bugcodes asm xdk)
 
+## GCC builds need this extra thing for some reason...
```
Comment (HBelusca) [inline boot/freeldr/freeldr/CMakeLists.txt]: Yes, if you remove that block GCC builds will complain. If you know a better way of doing this, please let me know.

File: boot/freeldr/freeldr/CMakeLists.txt [hunk 4]
```diff
@@ -233,8 +231,8 @@ if(MSVC)
         add_target_link_flags(freeldr_pe "/ignore:4078 /ignore:4254 /DRIVER")
         add_target_link_flags(freeldr_pe_dbg "/ignore:4078 /ignore:4254 /DRIVER")
     else()
-        add_target_link_flags(freeldr_pe "/ignore:4078 /ignore:4254 /DRIVER /FIXED /ALIGN:0x400 /SECTION:.text,ERW /SECTION:.data,RW /MERGE:.text16=.text /MERGE:INIT=.text /MERGE:.data=.text /MERGE:.rdata=.text /MERGE:.bss=.text")
-        add_target_link_flags(freeldr_pe_dbg "/ignore:4078 /ignore:4254 /DRIVER /FIXED /ALIGN:0x400 /SECTION:.text,ERW /SECTION:.data,RW /MERGE:.text16=.text /MERGE:INIT=.text /MERGE:.data=.text /MERGE:.rdata=.text /MERGE:.bss=.text")
+        add_target_link_flags(freeldr_pe "/ignore:4078 /ignore:4254 /DRIVER /FIXED /ALIGN:0x400 /SECTION:.text,ERW /SECTION:.data,RW /MERGE:INIT=.text /MERGE:.data=.text /MERGE:.rdata=.text /MERGE:.bss=.text")
+        add_target_link_flags(freeldr_pe_dbg "/ignore:4078 /ignore:4254 /DRIVER /FIXED /ALIGN:0x400 /SECTION:.text,ERW /SECTION:.data,RW /MERGE:INIT=.text /MERGE:.data=.text /MERGE:.rdata=.text /MERGE:.bss=.text")
```
Comment (HBelusca) [inline boot/freeldr/freeldr/CMakeLists.txt]: The `.text16` stuff doesn't exist anywhere else.

File: boot/freeldr/freeldr/arch/i386/drvmap.S [hunk 5]
```diff
@@ -40,14 +39,14 @@ Int13Handler:
     mov word ptr cs:[CallersFlags - Int13Handler], ax
 
     /* Save the drive number they passed in */
-    mov cs:[PassedInDriveNumber - Int13Handler], dl
+    mov byte ptr cs:[PassedInDriveNumber - Int13Handler], dl
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/drvmap.S]: In this file I'm basically adding the `<size> ptr` thingies explicitly.

File: boot/freeldr/freeldr/arch/i386/drvmap.S [hunk 6]
```diff
@@ -56,8 +55,8 @@ Int13Handler:
 
 GetMappedDriveNumberLoop:
 
-    /* Get the next drive number in the list */
-    lods ax, cs:[si]
+    /* Get the next drive number in the list (store it in AX) */
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/drvmap.S]: Mentioning the destination register seemed to be a GCC extension, but since AX is also the default in x86 for lods I just drop it and use the standard syntax that works in MSVC.

File: boot/freeldr/freeldr/arch/i386/drvmap.S [hunk 7]
```diff
@@ -139,3 +138,7 @@ Int13HandlerDriveNew4:
 
 PUBLIC _DriveMapInt13HandlerEnd
 _DriveMapInt13HandlerEnd:
+
+.endcode16
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/drvmap.S]: The .code16 needed its termination.

File: boot/freeldr/freeldr/arch/i386/entry.S [hunk 8]
```diff
@@ -25,6 +25,7 @@ EXTERN _BootMain:PROC
 EXTERN _InitIdt:PROC
 EXTERN _i386Idt:DWORD
 //EXTERN _i386idtptr:FWORD
+EXTERN cmdline:DWORD
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/entry.S]: `cmdline` is in multiboot.S

File: boot/freeldr/freeldr/arch/i386/entry.S [hunk 9]
```diff
@@ -73,27 +74,25 @@ _FrldrStartup:
     call _InitIdt
 
 #ifndef _USE_ML
-    /* Clean out bss */
+    /* Clean out BSS -- Only for GCC build since MSVC one has it merged with TEXT (see CMakeLists.txt) */
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/entry.S]: Recalling the reader why.

File: boot/freeldr/freeldr/arch/i386/mb.S [hunk 10]
```diff
@@ -30,9 +29,8 @@ PUBLIC _kernel_pagetable
 PUBLIC _apic_pagetable
 PUBLIC _PageDirectoryEnd
 
-#ifndef _USE_ML
 .bss
-#endif
+
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/mb.S]: That thing can compile with MASM; but it appears no definitions from there are actually used in FreeLdr at all.....
Leftover?

File: boot/freeldr/freeldr/arch/i386/multiboot.S [hunk 11]
```diff
@@ -98,12 +111,34 @@ mb3:
...
+
+    /* Get address of optional header */
+    mov eax, dword ptr es:[IMAGE_DOS_HEADER_e_lfanew]
+    add eax, 4 + IMAGE_FILE_HEADER_SIZE
+
+    /* Get the size of the image */
+    mov ecx, dword ptr es:[eax + IMAGE_OPTIONAL_HEADER_SizeOfImage]
+    // add ecx, FREELDR_PE_BASE - FREELDR_BASE /* == offset __bss_start__ - FREELDR_BASE since offset __bss_start__ == &__ImageBase (== FREELDR_PE_BASE) + OptionalHeader->SizeOfImage */
+    add ecx, offset ___ImageBase - FREELDR_BASE
+#endif
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/multiboot.S]: The idea of that block is that, since we don't know anything about BSS segment in the MSVC build I use some runtime way to guess the full image size of freeldr (including the 16-bit frldr16 stub that is prepended), based on the following code:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/lib/mm/meminit.c;hb=d27286655018c15688afe2dd88f1ac79dbc72aa9#l310

The code in comment is what I want to achieve.

Also I noticed that `__ImageBase` is actually equal to FREELDR_PE_BASE so I used it instead.

Comment (HBelusca) [reply boot/freeldr/freeldr/arch/i386/multiboot.S]: Any suggestions to improve that are welcome.

File: boot/freeldr/freeldr/arch/i386/multiboot.S [hunk 12]
```diff
@@ -98,12 +111,34 @@ mb3:
...
     shr ecx, 2
     rep movsd
 
     /* Load segment registers for real-address mode */
+#ifdef _USE_ML
+    lgdt fword ptr gdtptr
+#else
     lgdt gdtptr
+#endif
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/multiboot.S]: GCC vs. MASM (GCC doesn't like the explicit size specifier)

File: boot/freeldr/freeldr/arch/i386/multiboot.S [hunk 13]
```diff
@@ -122,7 +162,9 @@ mb4:
     mov cr0, eax
 
     /* Jump to real entry point */
-    ljmp16 0, FREELDR_BASE
+    ljmp 0, FREELDR_BASE
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/multiboot.S]: This needs to be verified. It was a local change originally.

File: boot/freeldr/freeldr/arch/i386/multiboot.S [hunk 14]
```diff
@@ -122,7 +162,9 @@ mb4:
...
+.endcode16
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/multiboot.S]: The .code16 has to be closed.

File: boot/freeldr/freeldr/arch/i386/multiboot.S [hunk 15]
```diff
@@ -139,10 +181,14 @@ gdt:
...
 
-    /* GDT pointer */
+/* GDT table pointer */
 gdtptr:
     .word HEX(17)       /* Limit */
+#ifndef _USE_ML
     .long gdt - 8       /* Base Address */
+#else
+    .long offset gdt - 8
+#endif
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/i386/multiboot.S]: If I don't use "offset", I get a wrong-size error from MASM/ML.

File: boot/freeldr/freeldr/arch/i386/multiboot.S (hunk 13 above)
Comment (tkreuzer) [inline boot/freeldr/freeldr/arch/i386/multiboot.S]: Why this change?

File: boot/freeldr/freeldr/arch/i386/multiboot.S [hunk 16]
```diff
@@ -139,10 +181,14 @@ gdt:
...
     .word HEX(0000)
 
-    /* GDT pointer */
+/* GDT table pointer */
 gdtptr:
     .word HEX(17)       /* Limit */
+#ifndef _USE_ML
     .long gdt - 8       /* Base Address */
+#else
+    .long offset gdt - 8
```
Comment (tkreuzer) [inline boot/freeldr/freeldr/arch/i386/multiboot.S]: Doesn't the latter work for both?

File: sdk/include/asm/asm.inc [hunk 17]
```diff
@@ -100,7 +100,7 @@ ENDM
 
 .code16 MACRO
     ASSUME nothing
-    .text SEGMENT use16
+    .text SEGMENT use16 PUBLIC 'CODE'
```
Comment (HBelusca) [inline sdk/include/asm/asm.inc]: That's the magic so that, when compiling and linking freeldr, we don't get *two separate* .text segments with different attributes, *even if* we ask all the .text/.data ... segments to be merged together in the freeldr linker command-line.

This avoids the `"v86.S.obj : warning LNK4078: multiple '.text' sections found with different attributes (C0520040)"` warning.

I have checked the generated code (freeldr_pe.exe) in IDA to see whether the incorporated 16-bit code was OK and it seemed to be so.

File: boot/freeldr/freeldr/arch/i386/multiboot.S [hunk 18]
```diff
@@ -98,12 +111,34 @@ mb3:
...
+#else
+    /* Address the image with es segment */
+    // mov ax, FREELDR_PE_BASE / 16 /* == &__ImageBase */
+    // mov es, ax
+    mov eax, offset ___ImageBase
+    shr eax, 4
+    mov es, eax
+
+    /* Get address of optional header */
+    mov eax, dword ptr es:[IMAGE_DOS_HEADER_e_lfanew]
```
Comment (tkreuzer) [inline boot/freeldr/freeldr/arch/i386/multiboot.S]: Why not much simpler:
```
lea eax, dword ptr cs:[___ImageBase]
add eax, dword ptr cs:[eax + IMAGE_DOS_HEADER_e_lfanew]
mov ecx, dword ptr cs:[eax + 4 + IMAGE_FILE_HEADER_SIZE + IMAGE_OPTIONAL_HEADER_SizeOfImage]
```

File: boot/freeldr/freeldr/CMakeLists.txt (hunk 3 above)
Comment (tkreuzer) [reply boot/freeldr/freeldr/CMakeLists.txt]: Someone invented that mini hal thing for whatever reasons. But IIRC it makes the MSVC build too large, so it's not used in MSVC builds, which also removes some extra features somewhere.

File: boot/freeldr/freeldr/arch/i386/multiboot.S (hunk 18 above)
Comment (HBelusca) [reply boot/freeldr/freeldr/arch/i386/multiboot.S]: I copy-pasted and adapted slightly the code from https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/realmode/i386.S;hb=d27286655018c15688afe2dd88f1ac79dbc72aa9#l36
(the same snippet also exists in amd64.S).
Your solution is more elegant, and may be ported back to the above-mentioned files too.

File: boot/freeldr/freeldr/arch/i386/multiboot.S (hunk 16 above)
Comment (HBelusca) [reply boot/freeldr/freeldr/arch/i386/multiboot.S]: I just retested; `.long gdt - 8` works in both ML and GAS.
On the contrary, `.long offset gdt - 8` works in ML but breaks GAS with the error:
```
{standard input}: Assembler messages:
{standard input}: Fatal error: Case value 61 unexpected at line 1152 of file "../../../../build/binutils/src/gas/symbols.c"
```

File: boot/freeldr/freeldr/arch/i386/multiboot.S (hunk 13 above)
Comment (HBelusca) [reply boot/freeldr/freeldr/arch/i386/multiboot.S]: I wanted to test why IDA output of the compiled module looked different in GCC vs. MSVC build,
but I've just realized that I did read the output incorrectly (if I ask IDA to reanalize this portion of code as 16-bit instructions I get the correct `jmp far ptr 0:0F800h ; nop;` (the nop is padding) instead of what I read while keeping 32-bit instructions: `jmp far ptr 8D90h:0F800h` ; the output from the MSVC build already contained a separate 16-bit section and so the instruction appeared correctly).

File: boot/freeldr/freeldr/CMakeLists.txt (hunk 2 above)
Comment (HBelusca) [reply boot/freeldr/freeldr/CMakeLists.txt]: Moving the asm up, seems to also put it mostly at the beginning of the freeldr_pe.exe, but yeah I agree it's purely decorative. I'll revert that.

File: sdk/include/asm/asm.inc (hunk 17 above)
Comment (Extravert-ir) [reply sdk/include/asm/asm.inc]: Maybe insert this as a comment?

File: boot/freeldr/freeldr/arch/i386/entry.S [hunk 19]
```diff
@@ -73,27 +74,25 @@ _FrldrStartup:
...
-#endif
 
     /* GO! */
     push eax
     call _BootMain
 
     /* We should never get here */
 stop:
-    jmp stop
+    jmp short stop
```
Comment (Extravert-ir) [inline boot/freeldr/freeldr/arch/i386/entry.S]: Aren't jumps "short" by default?

File: boot/freeldr/freeldr/arch/i386/drvmap.S (hunk 6 above)
Comment (Extravert-ir) [reply boot/freeldr/freeldr/arch/i386/drvmap.S]: It even can be just "lodsw"

File: boot/freeldr/freeldr/arch/i386/entry.S (hunk 19 above)
Comment (HBelusca) [reply boot/freeldr/freeldr/arch/i386/entry.S]: Yes indeed in that case it's assembled as "jmp short" even if "short" is not explicitly specified.

File: boot/freeldr/freeldr/arch/i386/entry.S (hunk 9 above)
Comment (tkreuzer) [reply boot/freeldr/freeldr/arch/i386/entry.S]: But it will most likely reduce the virtual size.

File: boot/freeldr/freeldr/arch/i386/mb.S (hunk 10 above)
Comment (tkreuzer) [reply boot/freeldr/freeldr/arch/i386/mb.S]: Does GCC build complain, when it's removed?

File: boot/freeldr/freeldr/arch/i386/multiboot.S (hunk 18 above)
Comment (tkreuzer) [reply boot/freeldr/freeldr/arch/i386/multiboot.S]: Real mode code cannot address more than 64k, so you need the segment addresing there.

File: boot/freeldr/freeldr/arch/i386/multiboot.S (hunk 11 above)
Comment (tkreuzer) [reply boot/freeldr/freeldr/arch/i386/multiboot.S]: Let's just merge it and get rid of this stuff entirely.

File: sdk/include/asm/asm.inc (hunk 17 above)
Comment (tkreuzer) [reply sdk/include/asm/asm.inc]: Nice! This might help to remove the freeldr16.bin hack at some point.

File: boot/freeldr/freeldr/CMakeLists.txt [hunk 20]
```diff
@@ -5,6 +5,74 @@ if(SEPARATE_DBG)
...
 
+if(NOT MSVC)
+###
+### For GCC
+###
+function(add_linker_script _target _linker_script_file)
+    get_filename_component(_file_full_path ${_linker_script_file} ABSOLUTE)
+    add_target_link_flags(${_target} "-Wl,-T,${_file_full_path}")
+
+    # Unfortunately LINK_DEPENDS is ignored in non-Makefile generators (for now...)
```
Comment (learn-more) [inline boot/freeldr/freeldr/CMakeLists.txt]: And this is an issue because?

Comment (HBelusca) [reply boot/freeldr/freeldr/CMakeLists.txt]: That's just a remark because according to CMake documentation, the LINK_DEPENDS property is enabled only for the "make" generator, and not for Ninja (in our old CMake version we are using). Newer CMake versions support the LINK_DEPENDS also for Ninja, see https://cmake.org/cmake/help/latest/prop_tgt/LINK_DEPENDS.html .

**Comment by HBelusca** [comment]
freeldr_msvc_multiboot_grub
